### ISSUE TITLE
Bad performance of CTAS if target is toast with redistribution motion.

### ISSUE THREAD

**Issue by paul-guo-** [greenplum-db/gpdb#9253]
### Greenplum version or build
Master

### Step to reproduce the behavior

```
CREATE OR REPLACE FUNCTION f_random_text( length integer)
RETURNS text AS
$body$
WITH chars AS (
    SELECT unnest(string_to_array('A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 0 1 2 3 4 5 6 7 8 9', ' ')) AS _char
),
charlist AS
(
    SELECT _char FROM chars ORDER BY random() LIMIT $1
)
SELECT string_agg(_char, '')
FROM charlist
;
$body$
LANGUAGE sql;

create table t14 (a text, b text, c text, d text, e text, f text, g text, h text, i text, j text, k text, l text, m text, n text, o text, p text, q text, r text, s text, t text, u text, v text, w text, x text, y text, z text);
insert into t14 select f_random_text(10), f_random_text(10), i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i, i from (select repeat('123456789', 10000) from generate_series(1,5000)) i;
analyze t14;
checkpoint;

\timing
create table tt as select * from t14 distributed by (b);
\timing
drop table tt; checkpoint;

\timing
create table tt as select a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y from t14 distributed by (b);
\timing
drop table tt; checkpoint;
```

Here are the running time. The running time gap between the two CTAS is super large. It's run on a mirrorless cluster. Note the 2nd CTAS just has one less column than the first one.
```
Time: 67237.978 ms
Time: 614.052 ms
```

Binary is compiled with -O0 but I've see similar big performance gap with -O3 before.

If we remove the `distributed by (b)` (that means no motion), the gap is much lower.
```
Time: 645.728 ms
Time: 540.539 ms
```

**Comment by hlinnaka** [comment]
I'm almost glad to see this, because I've been itching for some time to rewrite SerializeTuple function. It annoys me that it has two completely different ways of serializing a tuple: one for MemTuples and another for HeapTuples.

The cause of this issue is here, the HeapTuple codepath:
```

					/*
					 * If we have a toasted datum, forcibly detoast it here to
					 * avoid memory leakage: we want to force the detoast
					 * allocation(s) to happen in our reset-able serialization
					 * context.
					 */
					oldCtxt = MemoryContextSwitchTo(s_tupSerMemCtxt);
					attr = PointerGetDatum(PG_DETOAST_DATUM_PACKED(origattr));
					MemoryContextSwitchTo(oldCtxt);
```
That call to PG_DETOAST_DATUM_PACKED detoasts, *and decompresses*, the datum. The decompression is where all the CPU time is spent. To make things worse, in this case with CREATE TABLE AS, the datum will be compressed again in the receiver, so that it can be stored back in the table. So we're doing a completely pointless decomprsession+compression roundtrip, which is expensive.

The MemTuple codepath in SerializeTuple is smarter about that, and doesn't decompressed already-compresed datums. When you select all the columns in the table, the executor passes the HeapTuple directly from the table to the Motion, and you get into the slow HeapTuple codepath. But if there is any projection involved (or if the underlying table was an AO table), you get the fast MemTuple codepath.

Keeping my desire to rewrite the whole function in mind, I'm sad to say that this is very simple to fix without having to rewrite it :-). Opened a PR: https://github.com/greenplum-db/gpdb/pull/9256

**Comment by hlinnaka** [comment]
Fixed by PR #9256, closing.
